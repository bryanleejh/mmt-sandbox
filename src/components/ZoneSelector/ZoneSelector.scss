@import '~@deliveryhero/mmt-design-system/build/variables';

.zone-selector {
  display: flex;
  height: 330px;
  border: 1px solid $color-grey-200;
  border-radius: 0.25rem;

  .custom-control.custom-checkbox > * {
    cursor: pointer;
  }
}

.zone-panel {
  display: flex;
  flex-direction: column;
  height: 100%;
  min-width: 200px;
  max-width: 240px;
  width: 100%;
  border-right: 1px solid $color-grey-200;

  &__header {
    text-align: left;
    margin: 16px 0 8px;
  }

  &__header-text {
    font-size: 14px;
    font-weight: 500;
    height: 20px;
    letter-spacing: 0.3px;
    line-height: 20px;
    color: $color-brand-700;
    margin-bottom: 8px;
    padding-left: 16px;
    text-transform: capitalize;
  }

  &__header-input {
    display: flex;
    align-items: center;
    height: 32px;
    padding-left: 16px;

    & > .custom-control.custom-checkbox {
      transform: translateY(-2px);
    }

    &:hover {
      background-color: $color-active-on-white;
    }
  }

  &__header-label {
    text-transform: initial;
  }

  &__window {
    text-align: left;
    height: 100%;
    width: 100%;
  }

  &__item {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 0 8px 0 16px;
    cursor: pointer;

    & > .custom-control.custom-checkbox {
      transform: translateY(-2px);
      white-space: nowrap;
      overflow: hidden;
    }

    &--active {
      background-color: $color-active-on-white;
    }

    &:hover {
      background-color: $color-active-on-white;
    }
  }

  .custom-control-label {
    text-overflow: ellipsis;
  }
}

.zone-panel-fluid {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;

  svg.MuiSvgIcon-root.MuiSvgIcon-fontSizeSmall {
    display: none;
  }

  .zone-panel__item-text {
    min-width: none;
    max-width: none;
    width: 100%;
    text-overflow: initial;
  }
}