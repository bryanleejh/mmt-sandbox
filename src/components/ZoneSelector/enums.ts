export enum DisplayType {
  CITY = 'cities',
  REGION = 'regions',
  DISTRICT = 'districts',
}
