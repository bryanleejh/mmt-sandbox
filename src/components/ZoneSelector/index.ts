import {
  ZoneSelectorProps as ZoneSelectorPropsAlias,
  ZonePanelProps as ZonePanelPropsAlias,
  FormAreaState as FormAreaStateAlias,
} from './types';

export { default as ZoneSelector } from './ZoneSelector';

export type ZoneSelectorProps = ZoneSelectorPropsAlias;
export type ZonePanelProps = ZonePanelPropsAlias;
export type FormAreaState = FormAreaStateAlias;
