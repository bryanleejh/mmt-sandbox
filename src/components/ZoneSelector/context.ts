/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { DisplayType } from './enums';
import { IZoneContext } from './types';

export const defaultZoneContext = {
  formDisplayState: {
    [DisplayType.CITY]: [],
    [DisplayType.REGION]: [],
    [DisplayType.DISTRICT]: [],
  },
  formDisplayData: {
    [DisplayType.CITY]: [],
    [DisplayType.REGION]: [],
    [DisplayType.DISTRICT]: [],
  },
  formUpdate: {
    [DisplayType.CITY]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
    [DisplayType.REGION]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
    [DisplayType.DISTRICT]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
  },
  disabled: false,
};

export const ZoneContext = React.createContext<IZoneContext>(defaultZoneContext);

export const defaultZonePanelContext = {
  displayType: 'cities' as DisplayType,
};

export const ZonePanelContext = React.createContext(defaultZonePanelContext);
