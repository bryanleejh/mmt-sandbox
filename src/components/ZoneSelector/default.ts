/* eslint-disable @typescript-eslint/no-empty-function */
import { DisplayType } from './enums';

export const defaultZoneContext = {
  formDisplayState: {
    [DisplayType.CITY]: [],
    [DisplayType.REGION]: [],
    [DisplayType.DISTRICT]: [],
  },
  formDisplayData: {
    [DisplayType.CITY]: [],
    [DisplayType.REGION]: [],
    [DisplayType.DISTRICT]: [],
  },
  formUpdate: {
    [DisplayType.CITY]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
    [DisplayType.REGION]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
    [DisplayType.DISTRICT]: {
      onUpdateAll: () => {},
      onUpdate: () => {},
      onSelect: () => {},
    },
  },
  disabled: false,
};
