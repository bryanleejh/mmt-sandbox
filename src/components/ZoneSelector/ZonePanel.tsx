import React, { useRef, useContext, useEffect } from 'react';
import { Stack, Checkbox } from '@deliveryhero/armor';
import { equals, isEmpty, length, not, complement } from 'ramda';
import { FixedSizeList as List } from 'react-window';
import AutoSizer from 'react-virtualized-auto-sizer';
import ZonePanelItem from './ZonePanelItem';
import { ZoneContext } from './ZoneSelector';
import { hasIndeterState } from './helper';
import { DisplayType } from './enums';
import { ZonePanelProps } from './types';

export const ZonePanelContext = React.createContext({
  displayType: 'cities' as DisplayType,
});

const ZonePanel: React.FC<ZonePanelProps> = ({ displayType, itemData, fluid = false }: ZonePanelProps) => {
  const selectAllInputBox = useRef<any>(null);

  const { formDisplayState, formDisplayData, disabled } = useContext(ZoneContext);

  // const { onUpdateAll } = formUpdate[displayType];

  const isAllChecked = equals(length(formDisplayState[displayType]), length(formDisplayData[displayType]));

  const zonePanelContextObject = { displayType };

  useEffect(() => {
    if (selectAllInputBox.current) {
      const isFilled = complement(isEmpty);

      const indeterminateValue =
        (isFilled(formDisplayState[displayType]) && not(isAllChecked)) ||
        hasIndeterState(formDisplayState[displayType]);

      (selectAllInputBox.current as any).indeterminate = indeterminateValue;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectAllInputBox, formDisplayState, isAllChecked]);

  return (
    <div className={fluid ? 'zone-panel-fluid' : 'zone-panel'}>
      <div className='zone-panel__header'>
        <div className='zone-panel__header-text'>{displayType}</div>
        <div className='zone-panel__header-input'>
          <Stack ref={selectAllInputBox} id={`all-${displayType}`}>
            <Checkbox
              ref={selectAllInputBox}
              value={`all-${displayType}`}
              onChange={() => {}}
              checked={isAllChecked}
              data-testid={`all-${displayType}`}
              disabled={disabled}
              label={`Select all ${displayType}`}
            />
          </Stack>
        </div>
      </div>
      <div className='zone-panel__window'>
        <ZonePanelContext.Provider value={zonePanelContextObject}>
          <AutoSizer>
            {({ width, height }) => (
              <List itemData={itemData} itemCount={itemData.length} itemSize={32} height={height} width={width}>
                {ZonePanelItem}
              </List>
            )}
          </AutoSizer>
        </ZonePanelContext.Provider>
      </div>
    </div>
  );
};

export default ZonePanel;
