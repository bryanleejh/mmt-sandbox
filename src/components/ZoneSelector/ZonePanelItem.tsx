import React, { useRef, ChangeEvent, useEffect } from 'react';
// import { Bootstrap } from '@deliveryhero/mmt-design-system';
import { Stack, Checkbox } from '@deliveryhero/armor';
import { ListChildComponentProps } from 'react-window';
import { NavigateNext } from '@material-ui/icons';
import { paramCase } from 'param-case';
import theme from '../../styles/theme';
import { find, includes, propEq } from 'ramda';
import { ZoneContext } from './ZoneSelector';
import { ZonePanelContext } from './ZonePanel';
import { DisplayState } from './types';

// const { Form } = Bootstrap;

const ZonePanelItem = ({ data, index, style }: ListChildComponentProps) => {
  const selectInputBox = useRef<any>(null);
  const {
    formDisplayData,
    formDisplayState: { selectedCity, selectedRegion, ...formDisplayState },
    formUpdate,
    disabled,
  } = React.useContext(ZoneContext);
  const { displayType } = React.useContext(ZonePanelContext);
  const item = data[index];
  const id: string = item.slug || item.id;
  const isActive = includes(id, [selectedCity, selectedRegion]);
  const isZoneSelected = Boolean(find(propEq('id', id), formDisplayState[displayType]));

  const { onUpdate, onSelect } = formUpdate[displayType];

  useEffect(() => {
    if (selectInputBox.current) {
      const item = find<DisplayState>(propEq('id', id), formDisplayState[displayType]);

      (selectInputBox.current as any).indeterminate = item?.indeterminate;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectInputBox, formDisplayState, formDisplayData]);

  return (
    <div className={`zone-panel__item ${isActive ? 'zone-panel__item--active' : ''}`} style={style}>
      <Stack id={item.name}>
        <Checkbox
          ref={selectInputBox}
          value={item.slug}
          checked={isZoneSelected}
          disabled={disabled}
          data-testid={`checkbox-${paramCase(item.name)}`}
          onChange={() => {}}
          label={item.name}
        />
      </Stack>

      <NavigateNext
        data-testid={`select-${paramCase(item.name)}`}
        onClick={() => onSelect(item)}
        fontSize='small'
        style={{
          color: theme.colors.gray[300],
        }}
      />
    </div>
  );
};

export default ZonePanelItem;
