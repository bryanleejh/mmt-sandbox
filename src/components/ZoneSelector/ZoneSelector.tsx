import React, { useState, useEffect } from 'react';
import { isEmpty, map, filter, equals, length, clone, find, propEq, reject } from 'ramda';
import './ZoneSelector.scss';
import { Skeleton } from '@deliveryhero/armor';
import {
  Zone,
  City,
  ApiDisplayData,
  FormDisplayState,
  DisplayState,
  IZoneContext,
  ZoneSelectorProps,
  QueryResponse,
} from './types';
import { hasIndeterState, transformStateToUI, transformUIToState, updateIndeterState } from './helper';
import ZonePanel from './ZonePanel';
import { defaultZoneContext } from './default';
import { DisplayType } from './enums';
import { ApolloProvider, useLazyQuery } from '@apollo/react-hooks';
import { Country, getActiveCountry } from '../../utils/country';
import { GET_AREAS_QUERY, GET_CITIES_V2_QUERY } from '../../graphql/queries';
import { createApolloClient } from '../../graphql/client';

export const ZoneContext = React.createContext<IZoneContext>(defaultZoneContext);

const ZoneSelector: React.FC<ZoneSelectorProps> = ({
  selectedValues,
  onSelectValues,
  hasZones = false,
  disabled = false,
  overrideLoadCities,
  overrideLoadAreas,
}: ZoneSelectorProps) => {
  const [apiDisplayData, setApiDisplayData] = useState<ApiDisplayData>({
    cities: [],
    regions: [],
    districts: [],
  });
  const [formDisplayState, setFormDisplayState] = useState<FormDisplayState>({
    cities: [],
    regions: [],
    districts: [],
  });
  const [initialLoad, setInitialLoad] = useState(true);
  const [rawApiData, setRawApiData] = useState<QueryResponse[]>([]);
  const [selectedCity, setSelectedCity] = useState('');
  const [selectedRegion, setSelectedRegion] = useState('');

  const setCities = (citiesV2: City[]) => {
    setInitialLoad(false);
    if (isEmpty(apiDisplayData.cities)) {
      setApiDisplayData(state => ({
        ...state,
        cities: citiesV2,
      }));
    }
  };

  const setAreas = (areas: QueryResponse[]) => {
    setInitialLoad(false);
    if (!isEmpty(areas) && isEmpty(apiDisplayData.regions)) {
      setRawApiData(areas);
      const regions = map(area => area.region, areas);
      setApiDisplayData(state => ({
        ...state,
        regions,
      }));
    }
  };

  const [loadCities, { loading: cityLoading }] = useLazyQuery<{ citiesV2: City[] }, { country: Country }>(
    GET_CITIES_V2_QUERY,
    {
      variables: {
        country: getActiveCountry(),
      },
      onCompleted: ({ citiesV2 }) => {
        setCities(citiesV2);
      },
    }
  );

  const [loadAreas, { loading: areaLoading }] = useLazyQuery<
    { areas: QueryResponse[] },
    { country: Country; city: string }
  >(GET_AREAS_QUERY, {
    variables: {
      country: getActiveCountry(),
      city: selectedCity,
    },
    onCompleted: ({ areas }) => {
      setAreas(areas);
    },
  });

  useEffect(() => {
    const transformedState = transformStateToUI(selectedValues, selectedCity, selectedRegion, rawApiData);
    setFormDisplayState(transformedState);
  }, [selectedValues, apiDisplayData, selectedCity, selectedRegion, rawApiData]);

  useEffect(() => {
    if (isEmpty(apiDisplayData.cities)) {
      if (overrideLoadCities) {
        setInitialLoad(true);
        overrideLoadCities().then(citiesV2 => {
          if (!isEmpty(citiesV2)) {
            setCities(citiesV2);
          } else {
            setInitialLoad(false);
          }
        });
      } else {
        loadCities();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [apiDisplayData.cities]);

  // Fetch Regions
  useEffect(() => {
    if (isEmpty(apiDisplayData.regions) && selectedCity) {
      setInitialLoad(true);
      if (overrideLoadAreas) {
        overrideLoadAreas(selectedCity).then(areas => {
          setAreas(areas);
        });
      } else {
        loadAreas();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedCity, apiDisplayData.regions]);

  useEffect(() => {
    if (selectedRegion && rawApiData) {
      const [region] = filter(data => equals(data.region.id, selectedRegion), rawApiData);

      if (region && !isEmpty(region.districts)) {
        setApiDisplayData(prevState => ({
          ...prevState,
          districts: region.districts,
        }));
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedRegion, rawApiData]);

  const isLoading = cityLoading || areaLoading || initialLoad;

  const returnSelectedValues = (UIState: FormDisplayState) => {
    const newFormState = transformUIToState(
      {
        ...UIState,
        selectedCity,
        selectedRegion,
        isLoading,
        totalRegions: length(apiDisplayData.regions),
        totalDistricts: length(apiDisplayData.districts),
      },
      selectedValues
    );
    const isAllCitiesChecked = equals(apiDisplayData.cities.length, newFormState.length);
    const hasRegions = Boolean(newFormState.find(city => city.regions));
    const isAllCitiesSelected = isAllCitiesChecked && !hasRegions;
    onSelectValues({ isAllCitiesSelected, formState: newFormState });
  };

  // Mass Update Handlers
  const onUpdateAllDistricts = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    let cities = clone(formDisplayState.cities);
    let regions = clone(formDisplayState.regions);
    let districts = clone(formDisplayState.districts);
    const checkedRegion = { id: selectedRegion, indeterminate: false };

    if (checked) {
      const isRegionChecked = Boolean(find(propEq('id', selectedRegion), regions));
      if (!isRegionChecked) {
        regions.push(checkedRegion);
      }
      const isAllRegionsChecked = equals(length(regions), length(apiDisplayData.regions));

      const isCityChecked = Boolean(find(propEq('id', selectedCity), cities));

      if (isCityChecked) {
        cities = updateIndeterState(!isAllRegionsChecked, selectedCity, cities);
      } else {
        cities.push({ id: selectedCity, indeterminate: !isAllRegionsChecked });
      }

      districts = map(
        district => ({
          id: district.id,
          indeterminate: false,
        }),
        apiDisplayData.districts
      );

      regions = updateIndeterState(false, selectedRegion, regions);
    } else {
      districts = [];
      regions = isEmpty(formDisplayState.regions)
        ? formDisplayState.regions
        : reject(propEq('id', selectedRegion), regions);
      cities = isEmpty(regions)
        ? reject(propEq('id', selectedCity), cities)
        : updateIndeterState(true, selectedCity, cities);
    }

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  const onUpdateAllRegions = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    let cities = clone(formDisplayState.cities);
    let regions = clone(formDisplayState.regions);
    let districts = clone(formDisplayState.districts);

    if (checked) {
      if (isEmpty(cities)) {
        cities.push({ id: selectedCity, indeterminate: false });
      } else {
        cities = updateIndeterState(false, selectedCity, cities);
      }
      regions = map(
        region => ({
          id: region.id,
          indeterminate: false,
        }),
        apiDisplayData.regions
      );
      districts = map(
        district => ({
          id: district.id,
          indeterminate: false,
        }),
        apiDisplayData.districts
      );
    } else {
      cities = reject(propEq('id', selectedCity), cities);
      regions = [];
      districts = [];
    }

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  const onUpdateAllCities = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;

    const districts = checked
      ? map(
          district => ({
            id: district.id,
            indeterminate: false,
          }),
          apiDisplayData.districts
        )
      : [];

    const regions = checked
      ? map(
          region => ({
            id: region.id,
            indeterminate: false,
          }),
          apiDisplayData.regions
        )
      : [];

    const cities = checked
      ? map(
          city => ({
            id: city.slug,
            indeterminate: false,
          }),
          apiDisplayData.cities
        )
      : [];

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  // Single Update handlers
  const onUpdateDistrict = (e: React.ChangeEvent<HTMLInputElement>, item: Zone) => {
    const { checked } = e.target;
    const { id } = item;

    let districts = clone(formDisplayState.districts);
    let regions = clone(formDisplayState.regions);
    let cities = clone(formDisplayState.cities);

    const checkedDistrict: DisplayState = { id, indeterminate: false };
    const isRegionChecked = Boolean(find(propEq('id', selectedRegion), regions));
    const isCityChecked = Boolean(find(propEq('id', selectedCity), cities));

    if (checked) {
      districts.push(checkedDistrict);

      const isAllDistrictsChecked = equals(length(districts), length(apiDisplayData.districts));

      if (isRegionChecked) {
        regions = updateIndeterState(!isAllDistrictsChecked, selectedRegion, regions);
      } else {
        regions.push({
          id: selectedRegion,
          indeterminate: !isAllDistrictsChecked,
        });
      }

      const isAllRegionsChecked = equals(length(regions), length(apiDisplayData.regions));

      if (isCityChecked) {
        const indeterminateValue = hasIndeterState(regions) || !isAllRegionsChecked;
        cities = updateIndeterState(indeterminateValue, selectedCity, cities);
      } else {
        cities.push({ id: selectedCity, indeterminate: !isAllRegionsChecked });
      }
    } else {
      districts = reject(district => equals(district.id, id), districts);

      if (isEmpty(districts)) {
        regions = reject(region => equals(region.id, selectedRegion), regions);
      }

      if (!isEmpty(regions)) {
        regions = updateIndeterState(true, selectedRegion, regions);
      }

      if (isEmpty(regions)) {
        cities = reject(city => equals(city.id, selectedCity), cities);
      }

      if (!isEmpty(cities)) {
        cities = updateIndeterState(true, selectedCity, cities);
      }
    }

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  const onUpdateRegion = (e: React.ChangeEvent<HTMLInputElement>, item: Zone) => {
    const { checked } = e.target;
    const { id } = item;

    let districts = clone(formDisplayState.districts);
    let regions = clone(formDisplayState.regions);
    let cities = clone(formDisplayState.cities);

    const checkedRegion: DisplayState = { id, indeterminate: false };
    const isCityChecked = Boolean(find(propEq('id', selectedCity), cities));

    if (checked) {
      regions.push(checkedRegion);

      const isAllRegionsChecked = equals(length(regions), length(apiDisplayData.regions));

      if (isCityChecked) {
        cities = updateIndeterState(!isAllRegionsChecked, selectedCity, cities);
      } else {
        cities.push({ id: selectedCity, indeterminate: !isAllRegionsChecked });
      }

      // Check all districts if focused panel is current region
      if (equals(selectedRegion, id)) {
        districts = map(
          district => ({
            id: district.id,
            indeterminate: false,
          }),
          apiDisplayData.districts
        );
      }
    } else {
      regions = reject(region => equals(region.id, id), regions);

      // Uncheck all districts if focused panel is current region
      if (equals(selectedRegion, id)) {
        districts = [];
      }

      if (!isEmpty(regions)) {
        cities = updateIndeterState(true, selectedCity, cities);
      }

      if (isEmpty(regions)) {
        cities = reject(city => equals(city.id, selectedCity), cities);
      }
    }

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  const onUpdateCity = (e: React.ChangeEvent<HTMLInputElement>, item: City) => {
    const { checked } = e.target;
    const { slug } = item;

    let districts = clone(formDisplayState.districts);
    let regions = clone(formDisplayState.regions);
    let cities = clone(formDisplayState.cities);

    const checkedCity: DisplayState = { id: slug, indeterminate: false };

    if (checked) {
      cities.push(checkedCity);
      regions = map(
        region => ({
          id: region.id,
          indeterminate: false,
        }),
        apiDisplayData.regions
      );
      districts = map(
        district => ({
          id: district.id,
          indeterminate: false,
        }),
        apiDisplayData.districts
      );
    } else {
      cities = reject(city => equals(city.id, slug), cities);
      regions = [];
      districts = [];
    }

    returnSelectedValues({
      cities,
      regions,
      districts,
    });
  };

  const onCitySelect = (city: City) => {
    setSelectedCity(city.slug);
  };

  const onRegionSelect = (region: Zone) => {
    setSelectedRegion(region.id);
  };

  const zoneContextObject: IZoneContext = {
    formDisplayState: {
      [DisplayType.CITY]: formDisplayState.cities,
      [DisplayType.REGION]: formDisplayState.regions,
      [DisplayType.DISTRICT]: formDisplayState.districts,
      selectedCity,
      selectedRegion,
    },
    formDisplayData: apiDisplayData,
    formUpdate: {
      [DisplayType.CITY]: {
        onUpdateAll: onUpdateAllCities,
        onUpdate: onUpdateCity,
        onSelect: onCitySelect,
      },
      [DisplayType.REGION]: {
        onUpdateAll: onUpdateAllRegions,
        onUpdate: onUpdateRegion,
        onSelect: onRegionSelect,
      },
      [DisplayType.DISTRICT]: {
        onUpdateAll: onUpdateAllDistricts,
        onUpdate: onUpdateDistrict,
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        onSelect: () => {},
      },
    },
    disabled,
  };

  return (
    <ZoneContext.Provider value={zoneContextObject}>
      <div className='zone-selector'>
        {isLoading && <Skeleton />}
        <ZonePanel itemData={apiDisplayData.cities} displayType={DisplayType.CITY} fluid={!hasZones} />
        {!isEmpty(apiDisplayData.regions) && selectedCity && (
          <ZonePanel itemData={apiDisplayData.regions} displayType={DisplayType.REGION} />
        )}
        {!isEmpty(apiDisplayData.districts) && selectedRegion && (
          <ZonePanel itemData={apiDisplayData.districts} displayType={DisplayType.DISTRICT} fluid />
        )}
      </div>
    </ZoneContext.Provider>
  );
};

const WrappedZoneSelector: React.FC<ZoneSelectorProps> = props => {
  return (
    <ApolloProvider client={createApolloClient()}>
      <ZoneSelector {...props} />
    </ApolloProvider>
  );
};

export default WrappedZoneSelector;
