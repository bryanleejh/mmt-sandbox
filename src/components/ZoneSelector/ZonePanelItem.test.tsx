import React from 'react';
import ZonePanelItem from './ZonePanelItem';
import { render, fireEvent } from '@testing-library/react';
import { ZoneContext } from './ZoneSelector';
import { defaultZoneContext } from './default';
import { paramCase } from 'param-case';
import { ZonePanelContext } from './ZonePanel';
import { DisplayType } from './enums';

const mockData = [
  {
    id: 711,
    name: 'Singapore',
    slug: 'sg-singapore',
  },
];

const defaultComponentProps = {
  data: mockData,
  index: 0,
  style: {},
};

const customRender = (ui, { providerProps, ...renderOptions }) => {
  const mockPanelProvider = {
    value: {
      displayType: DisplayType.CITY,
    },
  };
  return render(
    <ZoneContext.Provider {...providerProps}>
      <ZonePanelContext.Provider {...mockPanelProvider}>{ui}</ZonePanelContext.Provider>
    </ZoneContext.Provider>,
    renderOptions
  );
};

describe('ZonePanelItem', () => {
  it('should render a single item', () => {
    const providerProps = {
      value: defaultZoneContext,
    };
    const { asFragment } = customRender(<ZonePanelItem {...defaultComponentProps} />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render disabled state', () => {
    const providerProps = {
      value: {
        ...defaultZoneContext,
        disabled: true,
      },
    };
    const { asFragment } = customRender(<ZonePanelItem {...defaultComponentProps} />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should check rendered checkbox based on display state', () => {
    const providerProps = {
      value: {
        ...defaultZoneContext,
        formDisplayState: {
          cities: [{ id: 'sg-singapore', indeterminate: true }],
        },
      },
    };
    const { getByTestId } = customRender(<ZonePanelItem {...defaultComponentProps} />, { providerProps });
    const panelItemCheckbox = getByTestId(`checkbox-${paramCase(mockData[0].name)}`);
    expect(panelItemCheckbox.checked).toEqual(true);
  });

  it('should render indeterminate state', () => {
    const providerProps = {
      value: {
        ...defaultZoneContext,
        formDisplayState: {
          cities: [{ id: 'sg-singapore', indeterminate: true }],
        },
      },
    };
    const { asFragment } = customRender(<ZonePanelItem {...defaultComponentProps} />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });
});
