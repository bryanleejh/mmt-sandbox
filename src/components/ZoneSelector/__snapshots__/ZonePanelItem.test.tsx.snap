// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZonePanelItem should render a single item 1`] = `
<DocumentFragment>
  <div
    class="zone-panel__item "
  >
    <div
      class="custom-control custom-checkbox"
    >
      <input
        class="custom-control-input"
        data-testid="checkbox-singapore"
        id="Singapore"
        type="checkbox"
        value="sg-singapore"
      />
      <label
        class="custom-control-label"
        for="Singapore"
      >
        Singapore
      </label>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
      data-testid="select-singapore"
      focusable="false"
      style="color: rgb(205, 206, 211);"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ZonePanelItem should render disabled state 1`] = `
<DocumentFragment>
  <div
    class="zone-panel__item "
  >
    <div
      class="custom-control custom-checkbox"
    >
      <input
        class="custom-control-input"
        data-testid="checkbox-singapore"
        disabled=""
        id="Singapore"
        type="checkbox"
        value="sg-singapore"
      />
      <label
        class="custom-control-label"
        for="Singapore"
      >
        Singapore
      </label>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
      data-testid="select-singapore"
      focusable="false"
      style="color: rgb(205, 206, 211);"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ZonePanelItem should render indeterminate state 1`] = `
<DocumentFragment>
  <div
    class="zone-panel__item "
  >
    <div
      class="custom-control custom-checkbox"
    >
      <input
        checked=""
        class="custom-control-input"
        data-testid="checkbox-singapore"
        id="Singapore"
        type="checkbox"
        value="sg-singapore"
      />
      <label
        class="custom-control-label"
        for="Singapore"
      >
        Singapore
      </label>
    </div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
      data-testid="select-singapore"
      focusable="false"
      style="color: rgb(205, 206, 211);"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
